embedded-compositor/qml/Keyboard.qml

```diff
@@ -32,37 +90,101 @@ Item {
 
     InputPanel {
         id: inputPanel
-        visible: y < parent.height
-        anchors.left: parent.left
-        anchors.right: parent.right
-        onYChanged: Qt.callLater(keyboardContainer.updateTranslate)
-        onHeightChanged: Qt.callLater(keyboardContainer.updateTranslate)
-        onVisibleChanged: keyboardContainer.updateTranslate()
-        state: inputPanel.active ? "VISIBLE": "HIDDEN"
+
+        width: keyboardContainer.width
+        height: keyboardContainer.height
+
+        rotation: keyboardContainer.rotation
+        state: "HIDDEN_0"
+
+        onYChanged: {
+            Qt.callLater(keyboardContainer.updateTranslate)
+        }
+
+        onXChanged: {
+            Qt.callLater(keyboardContainer.updateTranslate)
+        }
+
+        onHeightChanged: {
+            Qt.callLater(keyboardContainer.updateTranslate)
+        }
+
+        onWidthChanged: {
+            Qt.callLater(keyboardContainer.updateTranslate)
+        }
+
+        onVisibleChanged: {
+            keyboardContainer.updateTranslate()
+        }
 
         states: [
             State {
-                name: "VISIBLE"
+                name: "VISIBLE_0"
+                PropertyChanges {
+                    target: inputPanel
+                    y: keyboardContainer.height - inputPanel.height
+                }
+            },
+            State {
+                name: "HIDDEN_0"
+                PropertyChanges {
+                    target: inputPanel
+                    y: keyboardContainer.height
+                }
+            },
+            State {
+                name: "VISIBLE_90"
+                PropertyChanges {
+                    target: inputPanel
+                    x: keyboardContainer.width - inputPanel.width
+                }
+            },
+            State {
+                name: "HIDDEN_90"
+                PropertyChanges {
+                    target: inputPanel
+                    x: -keyboardContainer.height
```

Comment:
Shouldn't this be depend upon the width?
```suggestion
                    x: -keyboardContainer.width
```

Reply:
I think in the hidden case it does not really matter, it just has to be greater than the height of the keyboard.  So that it is not visible on the screen.